docs/networking/troubleshooting.md

```diff
@@ -0,0 +1,111 @@
+# Troubleshooting
+
+`coder ping <workspace>` will ping the workspace agent and print diagnostics on
+the state of the connection. These diagnostics are created by inspecting both
+the client and agent network configurations, and provide insights into why a
+direct connection may be impeded, or why the quality of one might be degraded.
+
+The `-v/--verbose` flag can be appended to the command to print client debug
+logs.
+
+```console
+$ coder ping dev
+pong from workspace proxied via  DERP(Council Bluffs, Iowa)  in 42ms
+pong from workspace proxied via  DERP(Council Bluffs, Iowa)  in 41ms
+pong from workspace proxied via  DERP(Council Bluffs, Iowa)  in 39ms
+✔ preferred DERP region: 999 (Council Bluffs, Iowa)
+✔ sent local data to Coder networking coordinator
+✔ received remote agent data from Coder networking coordinator
+    preferred DERP region: 999 (Council Bluffs, Iowa)
+    endpoints: x.x.x.x:46433, x.x.x.x:46433, x.x.x.x:46433
+✔ Wireguard handshake 11s ago
+
+❗ You are connected via a DERP relay, not directly (p2p)
+Possible client-side issues with direct connection:
+ - Network interface utun0 has MTU 1280, (less than 1378), which may degrade the quality of direct connections
+
+Possible agent-side issues with direct connection:
+ - Agent is potentially behind a hard NAT, as multiple endpoints were retrieved from different STUN servers
+ - Agent IP address is within an AWS range (AWS uses hard NAT)
+```
+
+## Common Problems
+
+### Direct Connections Disabled Deployment-Wide
+
+Direct connections can be disabled at the deployment level by setting the
+`CODER_BLOCK_DIRECT` environment variable or the `--block-direct-connections`
+flag on the server. When set, this will be reflected in the output of
+`coder ping`.
+
+### Firewall Blocks UDP
+
+Some corporate firewalls block UDP traffic. Direct connections require UDP
+traffic to be allowed between the client and agent, as well as between the
+client/agent and STUN servers in most cases. `coder ping` will indicate if
+either the Coder agent or client had issues sending or receiving UDP packets to
+STUN servers.
+
+If this is the case, you may need to add exceptions to the firewall to allow UDP
+for Coder workspaces, clients, and STUN servers.
+
+### Endpoint-Dependent NAT (Hard NAT)
+
+Hard NATs prevent public endpoints gathered from STUN servers from being used by
+the peer to establish a direct connection. Typically, if only one side of the
+connection is behind a hard NAT, direct connections can still be established
+easily. However, if both sides are behind hard NATs, direct connections may take
+longer to establish or may not be possible at all.
+
+Hard NATs may also be simulated by firewalls that block ingress traffic from
+unknown 5-tuples (Protocol, Source IP, Source Port, Destination IP, Destination
+Port).
+
+`coder ping` will indicate if it's possible the client or agent is behind a hard
+NAT or if a firewall is exhibiting this behavior.
```

Comment:
IIRC, we don't have any way to detect firewalls right now, so this isn't accurate.